velox/functions/sparksql/SIMDJsonFunctions.h

```diff
@@ -0,0 +1,195 @@
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig& config,
+      const arg_type<Varchar>* /*json*/,
+      const arg_type<Varchar>* jsonPath) {
+    if (jsonPath != nullptr) {
+      if (jsonPath->size() > 1 && jsonPath->data()[0] == '$') {
+        jsonPath_ = removeSingleQuotes(
+            std::string_view(jsonPath->data() + 1, jsonPath->size() - 1));
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& json,
+      const arg_type<Varchar>& jsonPath) {
+    // Spark requires the first char in jsonPath is '$'.
+    if (jsonPath.size() < 1 || jsonPath.data()[0] != '$') {
+      return false;
+    }
+    // jsonPath is "$".
+    if (jsonPath.size() == 1) {
+      result.append(json);
+      return true;
+    }
+    ParserContext ctx(json.data(), json.size());
+    ctx.parseDocument();
+    auto rawResult = jsonPath_.has_value()
+        ? ctx.jsonDoc.at_path(jsonPath_.value().data())
+        : ctx.jsonDoc.at_path(
+              removeSingleQuotes(
+                  std::string_view(jsonPath.data() + 1, jsonPath.size() - 1))
+                  .data());
+    if (rawResult.error()) {
+      return false;
+    }
+    auto error = extractStringResult(rawResult, result);
+    if (error) {
+      return false;
+    }
+
+    const char* currentPos;
+    ctx.jsonDoc.current_location().get(currentPos);
+    return isValidEndingCharacter(currentPos);
+  }
+
+ private:
+  // Spark's json path requires field name surrounded by single quotes if it is
+  // specified in "[]". But simdjson lib requires not. This method just removes
+  // such single quotes, e.g., converts "['a']['b']" to "[a][b]".
+  FOLLY_ALWAYS_INLINE std::string removeSingleQuotes(
+      std::string_view jsonPath) {
+    std::string result(jsonPath.data(), jsonPath.size());
+    size_t pairEnd = 0;
+    while (true) {
+      auto pairBegin = result.find("['", pairEnd);
+      if (pairBegin == std::string::npos) {
+        break;
+      }
+      pairEnd = result.find("]", pairBegin);
+      if (pairEnd == std::string::npos || result[pairEnd - 1] != '\'') {
+        return "-1";
+      }
+      result.erase(pairEnd - 1, 1);
+      result.erase(pairBegin + 1, 1);
+      pairEnd -= 2;
+    }
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE simdjson::error_code extractStringResult(
+      simdjson_result<ondemand::value> rawResult,
+      out_type<Varchar>& result) {
+    simdjson::error_code error;
+    std::stringstream ss;
+    switch (rawResult.type()) {
+      // For number and bool types, we need to explicitly get the value
+      // for specific types instead of using `ss << rawResult`. Thus, we
+      // can make simdjson's internal parsing position moved and then we
+      // can check the validity of ending character.
+      case ondemand::json_type::number: {
+        switch (rawResult.get_number_type()) {
```

Comment:
I wonder if this code repeats somewhere else. Would you check?

Reply:
Just checked again. I didn't find the repeat code anywhere else.